common/src/block_interval.rs

```diff
@@ -0,0 +1,255 @@
+    // acquired from the blockchain node.
+    FollowFrom(u64),
+}
+
+impl BlockInterval {
+    /// Create a new block interval
+    ///
+    /// A valid block range is of the form:
+    ///     * `block_number` for a single block number
+    ///     * `lhs..rhs`, `lhs..=rhs` as a exclusive/inclusive range
+    ///     * `lhs..` for a range starting from `lhs` to the chain tip. Form
+    ///       `lhs..=` is also valid format.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    ///    # use common::block_interval::BlockInterval;
+    ///    assert_eq!(BlockInterval::new("0..10").unwrap(), BlockInterval::Range(0..10));
+    ///    assert_eq!(BlockInterval::new("0..=10").unwrap(), BlockInterval::Range(0..11));
+    ///    assert_eq!(BlockInterval::new("32141").unwrap(), BlockInterval::Single(32141));
+    ///    assert_eq!(BlockInterval::new("100..").unwrap(), BlockInterval::FollowFrom(100));
+    /// ```
+    pub fn new(s: &str) -> Result<BlockInterval, BlockIntervalError> {
+        if (s.starts_with("0x") && s.len() == 66) || s.len() == 64 {
+            // Try to parse hash
+            let hash = s
+                .parse::<B256>()
+                .map_err(|_| BlockIntervalError::InvalidBlockHash(s.to_string()))?;
+            return Ok(BlockInterval::SingleBlockId(BlockId::Hash(hash.into())));
+        }
+
+        // First we parse for inclusive range and then for exclusive range,
+        // because both separators start with `..`
+        if let Ok(range) = parsing::parse_range_inclusive(s) {
+            Ok(BlockInterval::Range(range))
+        } else if let Ok(range) = parsing::parse_range_exclusive(s) {
+            Ok(BlockInterval::Range(range))
+        }
+        // Now we look for the follow from range
+        else if s.contains("..") {
+            let mut split = s.trim().split("..").filter(|s| *s != "=" && !s.is_empty());
+
+            // Any other character after `..` or `..=` is invalid
+            if split.clone().count() > 1 {
+                return Err(BlockIntervalError::InvalidFollowFromRange(s.to_string()));
+            }
+            let num = split
+                .next()
+                .map(|num| {
+                    num.parse::<u64>()
+                        .map_err(|_| BlockIntervalError::InvalidBlockNumber(s.to_string()))
+                })
+                .ok_or(BlockIntervalError::InvalidBlockIntervalRange(s.to_string()))??;
+            return Ok(BlockInterval::FollowFrom(num));
+        }
+        // Only single block number is left to try to parse
+        else {
+            let num = s
+                .trim()
+                .parse()
+                .map_err(|_| BlockIntervalError::InvalidBlockIntervalRange(s.to_string()))?;
+            return Ok(BlockInterval::Single(num));
+        }
+    }
+
+    /// Convert the block interval into an async stream of block numbers.
+    pub fn into_bounded_stream(self) -> Result<impl Stream<Item = u64>, BlockIntervalError> {
```

Comment:
Suppose I'm trying to handle this error, why should I have to deal with `BlockIntervalError::InvalidFollowFromRange`, that's a parse error but I _know_ I've already parsed successfully

Reply:
You will not deal with `InvalidFollowFromRange`. It is very common Rust practice to define error type for particular module comprising all the possible error states, and then deal only with the one that are relevant. If some function returns  `std::io::Error` you don't try to process/deal with all the possible I/O errors

Reply:
This is the equivalent of an operation that does no IO returning an `io::Error`

Reply:
It is not. `BlockIntervalError` communicates to the outside world all the possible errors that block interval handling could produce. Other modules using `BlockInterval` may wrap this error further.

Reply:
I don't think spotting a pattern and applying it is a good principle for error organisation.
Broadly, errors should be:
- informative to end users/library users.
- handleable, able to inform business logic.
- narrow, making illegal states unrepresentable.

_Different libraries make different tradeoffs according to usage._

I'll try and illustrate with [`std::sync::mpsc`](https://doc.rust-lang.org/std/sync/mpsc/index.html).
At the moment, it defines several _different_ error types.
- [`TryRecvError`](https://doc.rust-lang.org/std/sync/mpsc/enum.TryRecvError.html)
- [`TrySendError`](https://doc.rust-lang.org/std/sync/mpsc/enum.TrySendError.html)
- [`RecvTimeoutError`](https://doc.rust-lang.org/std/sync/mpsc/enum.RecvTimeoutError.html)
- [`RecvError`](https://doc.rust-lang.org/std/sync/mpsc/struct.RecvError.html)
- [`SendError`](https://doc.rust-lang.org/std/sync/mpsc/struct.SendError.html)

> It is very common Rust practice to define error type for particular module comprising all the possible error states, and then deal only with those that are relevant.

Why isn't `std` following this practice?
Well suppose they did have one giant `mpsc::Error`:
```rust
pub enum Error {
    Full
    Empty,
    Disconnected,
    Timeout
}
```
Then whenever I  _handle_ an error, here's what the code has to look like this:
```rust
match channel.try_recv() {
    Ok(_) => todo!(),
    Err(mpsc::Error::Timeout) => unreachable!("I didn't call recv_timeout, so I know this error can't be returned"),
    Err(_) => todo!(),
}
```
The _point_ is to not handle unapplicable variants.

But, as you observe, libraries like [`reqwest`](https://docs.rs/reqwest), _do_ define a top-level `reqwest::Error`, following the practice you alluded to. Why?

_because its intended use is like this:_
```rust
match request::get("http://example.com") {
    Ok(_) => todo!(),
    Err(_) => eprintln!("something went wrong fetching a web request"),
}
```

When I say this error enum should be removed, I'm saying that for this library:
- Parsing errors don't need to be so detailed - there is no user code that needs to distinguish between error variants.
  I suggest `FromStr<Err = anyhow::Error>` because it doesn't matter, though you may insist on a `ParseError<'static str>`.
- Once parsing is done, _a user should never need to worry about handling a parsing error_ - these are two error domains that must be separated.
  In fact, for your other methods I recommend returning an `Option` instead.

Reply:
Agree to disagree. I'll address this topic according to the team consensus - @cpubot , @Nashtare , @muursh what do you think - should we keep `BlockIntervalError` or just return string error/Option?

Reply:
I think I agree with @0xaatif here that there are several distinct error types here

Reply:
@muursh So do we keep couple of custom types or we switch to generic anyhow error string?

Reply:
Honestly I can't say I have a strong opinion there, so if someone else does I'm happy to go with whatever.

Reply:
Removed `BlockIntervalError` and used generic `anyhow::Error` everywhere